openhands/controller/agent_controller.py

```diff
@@ -234,8 +217,11 @@ async def on_event(self, event: Event) -> None:
 
         if isinstance(event, Action):
             await self._handle_action(event)
+            if event.source == EventSource.USER:
+                self.step()
         elif isinstance(event, Observation):
             await self._handle_observation(event)
+            self.step()
```

Comment:
Does this include obs like AgentChangeState or something which isn't supposed to interest the agent? It seems like it does

Reply:
Yeah good point. This triggering logic is the scariest part of the PR, I'll take another swing at it

Reply:
Ok I actually think we should be reacting to all Observations (agent state changed -> Running causes the first step, and the step function bails if not Running)

But I did make the logic more explicit here so it's easier to modify